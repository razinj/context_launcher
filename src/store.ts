// Redux
import { configureStore, combineReducers } from '@reduxjs/toolkit'
// Reducers
import appsListReducer from './slices/appsList'
import appsSearchReducer from './slices/appsSearch'
import recentAppsReducer from './slices/recentApps'
import FavoriteAppsReducer from './slices/favoriteApps'
import preferencesReducer from './slices/preferences'
// Storage
import AsyncStorage from '@react-native-async-storage/async-storage'
// Redux Persist
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
  createMigrate,
} from 'redux-persist'

const rootReducer = combineReducers({
  appsList: appsListReducer,
  appsSearch: appsSearchReducer,
  recentApps: recentAppsReducer,
  favoriteApps: FavoriteAppsReducer,
  preferences: preferencesReducer,
})

const migrations = {
  2: (state: any) => {
    return {
      ...state,
      preferences: {
        displayRecentApps: true,
        displayFavoriteApps: true,
      },
    }
  },
}

const persistConfig = {
  key: 'root',
  version: 2,
  storage: AsyncStorage,
  migrate: createMigrate(migrations, { debug: false }),
}

const persistedReducer = persistReducer(persistConfig, rootReducer)

export const store = configureStore({
  reducer: persistedReducer,
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }),
})

export const persistor = persistStore(store)

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
