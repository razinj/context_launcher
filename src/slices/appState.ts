import { createAction, createSlice, PayloadAction } from '@reduxjs/toolkit'
import { AppDetails } from '../models/app-details'
import { RenderedIn } from '../models/rendered-in'
import { RootState } from '../store'

interface AppsSearchState {
  query?: string
  result: AppDetails[]
}

export interface AppState {
  displayAllApps: boolean
  displaySettings: boolean
  menuAppDetails?: AppDetails
  displayAppMenu: boolean
  displaySortableFavoriteApps: boolean
  displaySortablePinnedApps: boolean
  displaySortableTemporaryPinnedApps: boolean
  search: AppsSearchState
}

const initialState: AppState = {
  displayAllApps: false,
  displaySettings: false,
  menuAppDetails: undefined,
  displayAppMenu: false,
  displaySortableFavoriteApps: false,
  displaySortablePinnedApps: false,
  displaySortableTemporaryPinnedApps: false,
  search: {
    query: undefined,
    result: [],
  },
}

export const appStateSlice = createSlice({
  name: 'appState',
  initialState,
  reducers: {
    // GlobalAppState
    setDisplayAllApps: (state: AppState, { payload }: PayloadAction<boolean>) => {
      state.displayAllApps = payload
    },
    setDisplaySettings: (state: AppState, { payload }: PayloadAction<boolean>) => {
      state.displaySettings = payload
    },
    setMenuAppDetails: (state: AppState, { payload }: PayloadAction<AppDetails | undefined>) => {
      state.menuAppDetails = payload
    },
    setDisplayAppMenu: (state: AppState, { payload }: PayloadAction<boolean>) => {
      state.displayAppMenu = payload
    },
    setDisplaySortableFavoriteApps: (state: AppState, { payload }: PayloadAction<boolean>) => {
      state.displaySortableFavoriteApps = payload
    },
    setDisplaySortablePinnedApps: (state: AppState, { payload }: PayloadAction<boolean>) => {
      state.displaySortablePinnedApps = payload
    },
    setDisplaySortableTemporaryPinnedApps: (state: AppState, { payload }: PayloadAction<boolean>) => {
      state.displaySortableTemporaryPinnedApps = payload
    },
    // AppsSearchState
    setAppsSearchQuery: (state: AppState, { payload }: PayloadAction<string | undefined>) => {
      state.search.query = payload
    },
    setAppsSearchResult: (state: AppState, { payload }: PayloadAction<AppDetails[]>) => {
      state.search.result = payload
    },
    resetAppsSearchState: (state: AppState) => {
      state.search.query = undefined
      state.search.result = []
    },
  },
})

export const {
  setDisplayAllApps,
  setDisplaySettings,
  setMenuAppDetails,
  setDisplayAppMenu,
  setDisplaySortableFavoriteApps,
  setDisplaySortablePinnedApps,
  setDisplaySortableTemporaryPinnedApps,
  setAppsSearchQuery,
  setAppsSearchResult,
  resetAppsSearchState,
} = appStateSlice.actions

export const toogleAllApps = createAction('appState/toogleAllApps')
export const appLaunch = createAction<{ renderedIn: RenderedIn; appDetails: AppDetails }>('appState/appLaunch')
export const appLaunchFromSearch = createAction('appState/appLaunchFromSearch')
export const sortFavoriteApps = createAction('appState/sortFavoriteApps')
export const sortPinnedApps = createAction('appState/sortPinnedApps')
export const sortTemporaryPinnedApps = createAction('appState/sortTemporaryPinnedApps')

export const selectAppState = (state: RootState) => state.appState
export const selectDisplayAllApps = (state: RootState) => state.appState.displayAllApps
export const selectDisplaySettings = (state: RootState) => state.appState.displaySettings
export const selectMenuAppDetails = (state: RootState) => state.appState.menuAppDetails
export const selectDisplayAppMenu = (state: RootState) => state.appState.displayAppMenu
export const selectDisplaySortableFavoriteApps = (state: RootState) => state.appState.displaySortableFavoriteApps
export const selectDisplaySortablePinnedApps = (state: RootState) => state.appState.displaySortablePinnedApps
export const selectDisplaySortableTemporaryPinnedApps = (state: RootState) => state.appState.displaySortableTemporaryPinnedApps
export const selectAppsSearchQuery = (state: RootState) => state.appState.search.query
export const selectAppsSearchResult = (state: RootState) => state.appState.search.result

export default appStateSlice.reducer
