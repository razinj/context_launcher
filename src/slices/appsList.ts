import { createAction, createSelector, createSlice, PayloadAction } from '@reduxjs/toolkit'
import { AppDetails } from '../models/app-details'
import { RootState } from '../store'
import { getAppsLetterIndex } from '../utils/alphabet-list'
import { displayToast } from '../utils/toast'

export interface AppsListState {
  list: AppDetails[]
}

const initialState: AppsListState = {
  list: [],
}

export const appsListSlice = createSlice({
  name: 'appsList',
  initialState,
  reducers: {
    setAppsList: (state: AppsListState, { payload }: PayloadAction<AppDetails[]>) => {
      // console.log('setAppsList called');
      displayToast(`apps found : ${payload.length}`)

      state.list = payload.sort((appOne: AppDetails, appTwo: AppDetails) => appOne.name.localeCompare(appTwo.name))
    },
  },
})

export const { setAppsList } = appsListSlice.actions

export const getAppsListAction = createAction('appsList/getAppsListAction')
export const appRemovedAction = createAction<string>('appsList/appRemovedAction')
export const revalidateAppsLists = createAction('appsList/revalidateAppsLists')

const selectAppsList = (state: RootState) => state.appsList.list

export const selectAppsListMemoized = createSelector(selectAppsList, (list: AppDetails[]) => list)
export const selectAppsLetterListMemoized = createSelector(selectAppsList, (list: AppDetails[]) =>
  getAppsLetterIndex(list)
)

export default appsListSlice.reducer
