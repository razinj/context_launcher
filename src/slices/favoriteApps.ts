// Redux
import { createSelector, createSlice, PayloadAction } from '@reduxjs/toolkit'
// Models
import { RootState } from '../store'
import { FavoriteApp } from '../models/favorite-app'

export interface FavoriteAppsState {
  list: FavoriteApp[]
}

const initialState: FavoriteAppsState = {
  list: [],
}

export const favoriteAppsSlice = createSlice({
  name: 'favoriteApps',
  initialState,
  reducers: {
    addFavoriteApp: (state: FavoriteAppsState, { payload }: PayloadAction<FavoriteApp>) => {
      if (state.list.length == 5) return

      const existingApp = state.list.find(({ name }: FavoriteApp) => name === payload.name)

      // Add to list only if it doesn't exists
      if (!existingApp) state.list.push({ ...payload })
    },
    removeFavoriteApp: (state: FavoriteAppsState, { payload }: PayloadAction<string>) => {
      const foundAppIndex = state.list.findIndex(({ name }: FavoriteApp) => name === payload)

      if (foundAppIndex === -1) return

      state.list.splice(foundAppIndex, 1)
    },
    setFavoriteApps: (state: FavoriteAppsState, { payload }: PayloadAction<FavoriteApp[]>) => {
      state.list = payload
    },
    clearFavoriteApps: (state: FavoriteAppsState) => {
      state.list = []
    },
  },
})

export const { addFavoriteApp, removeFavoriteApp, setFavoriteApps, clearFavoriteApps } = favoriteAppsSlice.actions

const selectFavoriteApps = (state: RootState) => state.favoriteApps.list

export const selectFavoriteAppsMemoized = createSelector(selectFavoriteApps, (list: FavoriteApp[]) => list)
export const selectFavoriteAppsCountMemoized = createSelector(selectFavoriteApps, (list: FavoriteApp[]) => list.length)

export default favoriteAppsSlice.reducer
