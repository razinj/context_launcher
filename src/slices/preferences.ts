// Redux
import { createSelector, createSlice, PayloadAction } from '@reduxjs/toolkit'
// Models
import { RootState } from '../store'

export interface PreferencesState {
  displayRecentApps: boolean
  displayFavoriteApps: boolean
}

const initialState: PreferencesState = {
  displayRecentApps: true,
  displayFavoriteApps: true,
}

export const preferencesSlice = createSlice({
  name: 'preferences',
  initialState,
  reducers: {
    displayRecentApps: (state: PreferencesState, { payload }: PayloadAction<boolean>) => {
      state.displayRecentApps = payload
    },
    displayFavoriteApps: (state: PreferencesState, { payload }: PayloadAction<boolean>) => {
      state.displayFavoriteApps = payload
    },
    resetPreferences: (state: PreferencesState) => {
      state.displayRecentApps = true
      state.displayFavoriteApps = true
    },
  },
})

export const { displayRecentApps, displayFavoriteApps, resetPreferences } = preferencesSlice.actions

const selectDisplayRecentApps = (state: RootState) => state.preferences.displayRecentApps
const selectDisplayFavoriteApps = (state: RootState) => state.preferences.displayFavoriteApps

export const selectDisplayRecentAppsMemoized = createSelector(selectDisplayRecentApps, (value: boolean) => value)
export const selectDisplayFavoriteAppsMemoized = createSelector(selectDisplayFavoriteApps, (value: boolean) => value)

export default preferencesSlice.reducer
