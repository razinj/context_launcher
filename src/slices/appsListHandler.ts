import { PayloadAction } from '@reduxjs/toolkit'
import { Channel, channel } from 'redux-saga'
import { call, fork, put, select, take, takeLatest } from 'redux-saga/effects'
import { AppDetails } from '../models/app-details'
import LauncherAppsModule from '../native-modules/LauncherAppsModule'
import { appRemovedAction, getAppsListAction, revalidateAppsLists, setAppsList } from './appsList'
import { removeFavoriteApp, removeFavoriteApps, selectFavoriteAppsMemoized } from './favoriteApps'
import {
  removePinnedApp,
  removePinnedApps,
  selectPinnedAppsMemoized,
  selectTemporaryPinnedAppsMemoized,
} from './pinnedApps'
import { removeRecentApp, removeRecentApps, selectRecentAppsMemoized } from './recentApps'

const notAvailable = 'NOT_AVAILABLE'

function* getAppsListActionHandler() {
  try {
    const applications: string = yield call(LauncherAppsModule.getApplications)
    const apps = JSON.parse(applications) as AppDetails[]
    yield put(setAppsList(apps))
  } catch (error: unknown) {
    yield put(setAppsList([]))
    console.error('Error getting apps list:', error)
  }
}

function* appRemovedActionHandler(action: PayloadAction<string>) {
  yield put(removeRecentApp(action.payload))
  const pinnedApp = { id: notAvailable, packageName: action.payload, name: notAvailable, icon: notAvailable }
  yield put(removePinnedApp({ app: pinnedApp, isPermanent: true })) // TODO: Check why I send an "app" inseatd of "packageName" for removal
  yield put(removePinnedApp({ app: pinnedApp, isPermanent: false }))
  yield put(removeFavoriteApp(action.payload))
}

// Revalidate all apps' lists
function* x() {
  console.log('x called')

  const applications: string = yield call(LauncherAppsModule.getApplications)
  const apps = JSON.parse(applications) as AppDetails[]
  yield put(setAppsList(apps))

  const appsPackageNames = apps.map(({ packageName }: AppDetails) => packageName)

  // Revalidate recent apps
  const recentApps: AppDetails[] = yield select(selectRecentAppsMemoized)
  yield put(
    removeRecentApps(
      recentApps
        .map(({ packageName }: AppDetails) => packageName)
        .filter(
          (packageName: string) => !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
        )
    )
  )

  // Revalidate favorite apps
  const favoriteApps: AppDetails[] = yield select(selectFavoriteAppsMemoized)
  yield put(
    removeFavoriteApps(
      favoriteApps
        .map(({ packageName }: AppDetails) => packageName)
        .filter(
          (packageName: string) => !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
        )
    )
  )

  // Revalidate pinned apps
  const pinnedApps: AppDetails[] = yield select(selectPinnedAppsMemoized)
  yield put(
    removePinnedApps({
      isPermanent: true,
      packageNames: pinnedApps
        .map(({ packageName }: AppDetails) => packageName)
        .filter(
          (packageName: string) => !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
        ),
    })
  )

  // Revalidate temporary pinned apps
  const temporaryPinnedApps: AppDetails[] = yield select(selectTemporaryPinnedAppsMemoized)
  yield put(
    removePinnedApps({
      isPermanent: false,
      packageNames: temporaryPinnedApps
        .map(({ packageName }: AppDetails) => packageName)
        .filter(
          (packageName: string) => !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
        ),
    })
  )
}

// const upsertAppsListChannel = channel()

// function* watchAppsListUpsert() {
//   while (true) {
//     type x = typeof setAppsList
//     const action: x = yield take(upsertAppsListChannel)
//     yield put(action)
//   }
// }

function getApplicationsV2Async() {
  return new Promise((resolve, reject) => {
    LauncherAppsModule.getApplicationsV2((applications: string) => {
      resolve(applications)
    })
  })
}

function* y() {
  console.log('y called')

  const applications: string = yield call(getApplicationsV2Async)
  yield call(handleApplications, applications)

  const apps = JSON.parse(applications) as AppDetails[]

  const appsPackageNames = apps.map(({ packageName }: AppDetails) => packageName)
  // const appsPackageNames: string[] = []

  // Revalidate recent apps
  const recentApps: AppDetails[] = yield select(selectRecentAppsMemoized)
  yield put(
    removeRecentApps(
      recentApps
        .map(({ packageName }: AppDetails) => packageName)
        .filter(
          (packageName: string) => !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
        )
    )
  )

  // Revalidate favorite apps
  const favoriteApps: AppDetails[] = yield select(selectFavoriteAppsMemoized)
  yield put(
    removeFavoriteApps(
      favoriteApps
        .map(({ packageName }: AppDetails) => packageName)
        .filter(
          (packageName: string) => !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
        )
    )
  )

  // Revalidate pinned apps
  const pinnedApps: AppDetails[] = yield select(selectPinnedAppsMemoized)
  yield put(
    removePinnedApps({
      isPermanent: true,
      packageNames: pinnedApps
        .map(({ packageName }: AppDetails) => packageName)
        .filter(
          (packageName: string) => !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
        ),
    })
  )

  // Revalidate temporary pinned apps
  const temporaryPinnedApps: AppDetails[] = yield select(selectTemporaryPinnedAppsMemoized)
  yield put(
    removePinnedApps({
      isPermanent: false,
      packageNames: temporaryPinnedApps
        .map(({ packageName }: AppDetails) => packageName)
        .filter(
          (packageName: string) => !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
        ),
    })
  )
}

function* handleApplications(applications: string) {
  console.log('handleApplications called')

  const apps = JSON.parse(applications) as AppDetails[]
  // dispatch an action here
  yield put(setAppsList(apps))
}

function* childSaga(channel: Channel<any>) {
  const applications: string = yield call(LauncherAppsModule.getApplications)
  const apps = JSON.parse(applications) as AppDetails[]
  // yield put(setAppsList(apps))
  // const result = yield call(someAsyncFunction);
  yield put(channel, apps) // Emit the result through the channel
}

function* z() {
  console.log('z called')

  const communicationChannel: Channel<AppDetails[]> = yield call(channel)
  yield fork(childSaga, communicationChannel)

  while (true) {
    const result: AppDetails[] = yield take(communicationChannel)
    console.log(' result: ', result.length)

    const appsPackageNames = result.map(({ packageName }: AppDetails) => packageName)
    // const appsPackageNames: string[] = []

    // Revalidate recent apps
    const recentApps: AppDetails[] = yield select(selectRecentAppsMemoized)
    yield put(
      removeRecentApps(
        recentApps
          .map(({ packageName }: AppDetails) => packageName)
          .filter(
            (packageName: string) =>
              !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
          )
      )
    )

    // Revalidate favorite apps
    const favoriteApps: AppDetails[] = yield select(selectFavoriteAppsMemoized)
    yield put(
      removeFavoriteApps(
        favoriteApps
          .map(({ packageName }: AppDetails) => packageName)
          .filter(
            (packageName: string) =>
              !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
          )
      )
    )

    // Revalidate pinned apps
    const pinnedApps: AppDetails[] = yield select(selectPinnedAppsMemoized)
    yield put(
      removePinnedApps({
        isPermanent: true,
        packageNames: pinnedApps
          .map(({ packageName }: AppDetails) => packageName)
          .filter(
            (packageName: string) =>
              !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
          ),
      })
    )

    // Revalidate temporary pinned apps
    const temporaryPinnedApps: AppDetails[] = yield select(selectTemporaryPinnedAppsMemoized)
    yield put(
      removePinnedApps({
        isPermanent: false,
        packageNames: temporaryPinnedApps
          .map(({ packageName }: AppDetails) => packageName)
          .filter(
            (packageName: string) =>
              !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
          ),
      })
    )

    // Use the result returned by the child saga
    // Non-blocking logic here

    // ...
  }

  console.log('done')
}

export function* appsListSagas() {
  yield takeLatest(getAppsListAction.type, getAppsListActionHandler)
  yield takeLatest(appRemovedAction.type, appRemovedActionHandler)
  // yield takeLatest(revalidateAppsLists.type, x)
  // yield takeLatest(revalidateAppsLists.type, z)
  yield takeLatest(revalidateAppsLists.type, z)
}
