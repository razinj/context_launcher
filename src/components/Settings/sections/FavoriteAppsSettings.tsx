import React, { useMemo } from 'react'
import { Pressable, Switch, View } from 'react-native'
import { useDispatch, useSelector } from 'react-redux'
import { sortFavoriteApps } from '../../../slices/appState'
import { clearFavoriteApps, selectFavoriteAppsCountMemoized } from '../../../slices/favoriteApps'
import { displayFavoriteApps, selectDisplayFavoriteAppsMemoized } from '../../../slices/preferences'
import { displayToast } from '../../../utils/toast'
import SettingsItemLabel from '../shared/SettingsItemLabel'
import {
  activeSwitch,
  inActiveSwitch,
  settingItemButtonRippleConfig,
  settingItemWrapperStyle,
  settingsPressableItemStyle,
  switchTrackColor,
} from '../shared/values'

const FavoriteAppsSettings = () => {
  const dispatch = useDispatch()
  const favoriteAppsCount = useSelector(selectFavoriteAppsCountMemoized)
  const displayFavoriteAppsValue = useSelector(selectDisplayFavoriteAppsMemoized)

  const toggleDisplayFavoriteApps = () => {
    dispatch(displayFavoriteApps(!displayFavoriteAppsValue))
  }

  const onFavoriteAppsSortViewClick = () => {
    dispatch(sortFavoriteApps())
  }

  const onClearFavoriteApps = () => {
    dispatch(clearFavoriteApps())
    displayToast('Favorite apps cleared successfully!')
  }

  const favoriteAppsSortDisabled = useMemo(
    () => favoriteAppsCount <= 1 || !displayFavoriteAppsValue,
    [favoriteAppsCount, displayFavoriteAppsValue]
  )

  return (
    <>
      <View style={settingItemWrapperStyle}>
        <SettingsItemLabel title='Display' />
        <Switch
          testID='display-switch'
          value={displayFavoriteAppsValue}
          onValueChange={toggleDisplayFavoriteApps}
          trackColor={switchTrackColor}
          thumbColor={displayFavoriteAppsValue ? activeSwitch : inActiveSwitch}
        />
      </View>

      <View style={settingItemWrapperStyle}>
        <Pressable
          testID='sort-button'
          disabled={favoriteAppsSortDisabled}
          onPress={onFavoriteAppsSortViewClick}
          android_disableSound={true}
          android_ripple={settingItemButtonRippleConfig}
          style={[settingsPressableItemStyle, { opacity: favoriteAppsSortDisabled ? 0.5 : 1 }]}>
          <SettingsItemLabel
            title='Sort'
            description={
              favoriteAppsSortDisabled
                ? displayFavoriteAppsValue
                  ? 'Add more apps to be able to sort'
                  : 'Display favorite apps to be able to sort'
                : 'Click to start sorting'
            }
          />
        </Pressable>
      </View>

      <View style={settingItemWrapperStyle}>
        <Pressable
          testID='clear-button'
          onPress={onClearFavoriteApps}
          android_disableSound={true}
          android_ripple={settingItemButtonRippleConfig}
          style={settingsPressableItemStyle}>
          <SettingsItemLabel title='Clear' />
        </Pressable>
      </View>
    </>
  )
}

export default FavoriteAppsSettings
