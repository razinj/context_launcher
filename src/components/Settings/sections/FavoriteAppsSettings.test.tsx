import { fireEvent, screen } from '@testing-library/react-native'
import React from 'react'
import { ToastAndroid } from 'react-native'
import { initialStoreState } from '../../../../utils/test/data'
import { renderWithProvider } from '../../../../utils/test/utils'
import { sortFavoriteApps } from '../../../slices/appState'
import { clearFavoriteApps } from '../../../slices/favoriteApps'
import { displayFavoriteApps } from '../../../slices/preferences'
import { RootState } from '../../../store'
import FavoriteAppsSettings from './FavoriteAppsSettings'

const useDispatchMock = jest.fn()

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => useDispatchMock,
}))

describe('<FavoriteAppsSettings /> Tests', () => {
  beforeEach(() => {
    jest.spyOn(ToastAndroid, 'show')
  })

  it('should render correctly and match snapshot', () => {
    renderWithProvider(<FavoriteAppsSettings />)

    expect(screen.toJSON()).toMatchSnapshot()
    expect(screen.getByTestId('display-favorite-apps-switch')).toBeOnTheScreen()
    expect(screen.getByTestId('sort-favorite-apps-button')).toBeOnTheScreen()
    expect(screen.getByTestId('clear-favorite-apps-button')).toBeOnTheScreen()
  })

  it('should dispatch action to clear favorite apps when button is pressed', () => {
    const customInitialState: RootState = {
      ...initialStoreState,
      favoriteApps: {
        list: [
          {
            packageName: 'com.google.chrome',
            name: 'Chrome',
            icon: 'ICON',
          },
          {
            packageName: 'com.google.maps',
            name: 'Maps',
            icon: 'ICON',
          },
        ],
      },
    }

    renderWithProvider(<FavoriteAppsSettings />, { preloadedState: customInitialState })

    const clearFavoriteAppsButton = screen.getByTestId('clear-favorite-apps-button')

    expect(clearFavoriteAppsButton).toBeOnTheScreen()

    fireEvent.press(clearFavoriteAppsButton)

    expect(useDispatchMock).toBeCalledWith(clearFavoriteApps())
    expect(ToastAndroid.show).toBeCalledWith('Favorite apps cleared successfully!', ToastAndroid.LONG)
  })

  describe('Sort favorite apps button tests', () => {
    it('should render sort favorite apps button as disabled if favorite apps are <= 1', () => {
      const customInitialStoreState: RootState = {
        ...initialStoreState,
        favoriteApps: {
          list: [
            {
              packageName: 'com.google.chrome',
              name: 'Chrome',
              icon: 'ICON',
            },
          ],
        },
        preferences: {
          ...initialStoreState.preferences,
          displayFavoriteApps: true,
        },
      }

      renderWithProvider(<FavoriteAppsSettings />, { preloadedState: customInitialStoreState })

      expect(screen.getByTestId('sort-favorite-apps-button')).toBeDisabled()
      expect(screen.getByText(/Add more apps to be able to sort/)).toBeOnTheScreen()
    })

    it('should render sort favorite apps button as disabled if display favorite apps value is false', () => {
      const customInitialStoreState: RootState = {
        ...initialStoreState,
        favoriteApps: {
          list: [
            {
              packageName: 'com.google.chrome',
              name: 'Chrome',
              icon: 'ICON',
            },
            {
              packageName: 'com.google.maps',
              name: 'Maps',
              icon: 'ICON',
            },
          ],
        },
        preferences: {
          ...initialStoreState.preferences,
          displayFavoriteApps: false,
        },
      }

      renderWithProvider(<FavoriteAppsSettings />, { preloadedState: customInitialStoreState })

      expect(screen.getByTestId('sort-favorite-apps-button')).toBeDisabled()
      expect(screen.getByText(/Display favorite apps to be able to sort/)).toBeOnTheScreen()
    })

    it('should render sort favorite apps button as enabled if favorite apps are > 1 and display favorite apps value is true', () => {
      const customInitialStoreState: RootState = {
        ...initialStoreState,
        favoriteApps: {
          list: [
            {
              packageName: 'com.google.chrome',
              name: 'Chrome',
              icon: 'ICON',
            },
            {
              packageName: 'com.google.maps',
              name: 'Maps',
              icon: 'ICON',
            },
          ],
        },
        preferences: {
          ...initialStoreState.preferences,
          displayFavoriteApps: true,
        },
      }

      renderWithProvider(<FavoriteAppsSettings />, { preloadedState: customInitialStoreState })

      expect(screen.getByTestId('sort-favorite-apps-button')).not.toBeDisabled()
      expect(screen.getByText(/Click to start sorting/)).toBeOnTheScreen()
    })

    it('should dispatch action to sort favorite when button is pressed', () => {
      const customInitialStoreState: RootState = {
        ...initialStoreState,
        favoriteApps: {
          list: [
            {
              packageName: 'com.google.chrome',
              name: 'Chrome',
              icon: 'ICON',
            },
            {
              packageName: 'com.google.maps',
              name: 'Maps',
              icon: 'ICON',
            },
          ],
        },
        preferences: {
          ...initialStoreState.preferences,
          displayFavoriteApps: true,
        },
      }

      renderWithProvider(<FavoriteAppsSettings />, { preloadedState: customInitialStoreState })

      const sortFavoriteAppsButton = screen.getByTestId('sort-favorite-apps-button')

      expect(sortFavoriteAppsButton).toBeOnTheScreen()
      expect(sortFavoriteAppsButton).not.toBeDisabled()

      fireEvent.press(sortFavoriteAppsButton)

      expect(useDispatchMock).toBeCalledWith(sortFavoriteApps())
    })
  })

  it('should dispatch action to toggle display favorite apps when pressed', () => {
    const customInitialStoreState: RootState = {
      ...initialStoreState,
      preferences: {
        ...initialStoreState.preferences,
        displayFavoriteApps: false,
      },
    }

    renderWithProvider(<FavoriteAppsSettings />, { preloadedState: customInitialStoreState })

    const toggleFavoriteAppsSwitch = screen.getByTestId('display-favorite-apps-switch')

    expect(toggleFavoriteAppsSwitch).toBeOnTheScreen()
    expect(toggleFavoriteAppsSwitch).not.toBeDisabled()

    fireEvent(toggleFavoriteAppsSwitch, 'valueChange')

    expect(useDispatchMock).toBeCalledWith(displayFavoriteApps(true))
  })
})
