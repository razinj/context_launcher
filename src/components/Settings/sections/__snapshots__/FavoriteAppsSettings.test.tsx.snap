// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FavoriteAppsSettings /> Tests should render correctly and match snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        style={
          [
            {
              "padding": 10,
            },
            undefined,
          ]
        }
        testID="wrapper"
      >
        <Text
          style={
            [
              {
                "color": "#808080",
              },
              undefined,
            ]
          }
          testID="title"
        >
          Display
        </Text>
      </View>
      <RCTSwitch
        accessibilityRole="switch"
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onTintColor="#e5e5e5"
        style={
          {
            "height": 31,
            "width": 51,
          }
        }
        testID="display-switch"
        thumbTintColor="#42855B"
        tintColor="#e5e5e5"
        value={true}
      />
    </View>
    <View
      style={
        {
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": true,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
            },
            {
              "opacity": 0.5,
            },
          ]
        }
        testID="sort-button"
      >
        <View
          style={
            [
              {
                "padding": 10,
              },
              undefined,
            ]
          }
          testID="wrapper"
        >
          <Text
            style={
              [
                {
                  "color": "#808080",
                },
                undefined,
              ]
            }
            testID="title"
          >
            Sort
          </Text>
          <Text
            style={
              {
                "color": "#ccc",
              }
            }
          >
            Add more apps to be able to sort
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
        testID="clear-button"
      >
        <View
          style={
            [
              {
                "padding": 10,
              },
              undefined,
            ]
          }
          testID="wrapper"
        >
          <Text
            style={
              [
                {
                  "color": "#808080",
                },
                undefined,
              ]
            }
            testID="title"
          >
            Clear
          </Text>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
